libsolidity/interface/CompilerStack.cpp

```diff
@@ -103,6 +107,101 @@ CompilerStack::~CompilerStack()
 	TypeProvider::reset();
 }
 
+void CompilerStack::createAndAssignCallGraphs()
+{
+	for (Source const* source: m_sourceOrder)
+	{
+		if (!source->ast)
+			continue;
+
+		for (ContractDefinition const* contract: ASTNode::filteredNodes<ContractDefinition>(source->ast->nodes()))
+
+		{
+			ContractDefinitionAnnotation& annotation =
+				m_contracts.at(contract->fullyQualifiedName()).contract->annotation();
+
+			annotation.creationCallGraph = make_unique<CallGraph>(
+				FunctionCallGraphBuilder::buildCreationGraph(
+					*contract
+				)
+			);
+			annotation.deployedCallGraph = make_unique<CallGraph>(
+				FunctionCallGraphBuilder::buildDeployedGraph(
+					*contract,
+					**annotation.creationCallGraph
+				)
+			);
+
+			solAssert(annotation.contractDependencies.empty(), "contractDependencies expected to be empty?!");
+
+			annotation.contractDependencies = annotation.creationCallGraph->get()->bytecodeDependency;
+
+			for (auto const& [dependencyContract, referencee]: annotation.deployedCallGraph->get()->bytecodeDependency)
+				annotation.contractDependencies.emplace(dependencyContract, referencee);
+		}
+	}
+}
+
+void CompilerStack::findAndReportCyclicContractDependencies()
+{
+	// Cycles we found, used to avoid duplicate reports for the same reference
+	set<ASTNode const*, ASTNode::CompareByID> foundCycles;
+
+	for (Source const* source: m_sourceOrder)
+	{
+		if (!source->ast)
+			continue;
+
+		for (ContractDefinition const* contractDefinition: ASTNode::filteredNodes<ContractDefinition>(source->ast->nodes()))
+		{
+			// Ignore libraries as they are tested as part of the
+			// contracts that use them.
+			if (contractDefinition->isLibrary())
+				continue;
+
+			util::CycleDetector<ContractDefinition> cycleDetector{[&](
+				ContractDefinition const& _contract,
+				util::CycleDetector<ContractDefinition>& _cycleDetector,
+				size_t _depth
+			)
+			{
+				if (_depth >= MaxCycleDetectionRecursionDepth)
+					m_errorReporter.fatalTypeError(
+						7864_error,
+						_contract.location(),
+						"Contract dependencies exhausting cyclic dependency validator"
+					);
+
+				for (auto& [dependencyContract, referencee]: _contract.annotation().contractDependencies)
+					if (_cycleDetector.run(*dependencyContract))
+						return;
+			}};
+
+			ContractDefinition const* cycle = cycleDetector.run(*contractDefinition);
+
+			if (!cycle)
+				continue;
+
+			ASTNode const* referencee = contractDefinition->annotation().contractDependencies.at(cycle);
```

Comment:
What if we have a cycle of the form
`a -> b -> c -> d -> c`?
Then the return vaule of the `cycleDetector` will be `c`, which is not a dependency of `a`

Reply:
This is okay (I also tested your example just now) as the cycle Detector returns only and always the vertex that has a distance of 1 (depth = 1) to the originating vertex

Reply:
Interesting, this is not how CycleDetector should work. I will double-check, because I thought it works correctly with constants.

Reply:
Yeah, the code in line 64 of Algorithms.h does that. Ok, let's just leave it like that.